### DISCUSSION TITLE
Mobile App Flow

### DISCUSSION THREAD

**Discussion by MTRNord** [keycloak/keycloak#22530] [Q&A] [answered]
Hi,

I am currently building mobile apps and was wondering what the best user flow is from keycloak's pov.

I can obviously easily login using an OAuth2 client with a native UI. That's the easy part.
However, for registration, I fail to see how I can use the native UI for that. Does someone have an idea or hint to prevent the need of leaving the app? (Self Registration is obviously enabled on the realm)

**Comment by justin-tay** [comment, marked as answer]
The best practice for OAuth2 for the mobile flow for a native mobile application is detailed in [RFC 8252 - OAuth 2.0 for Native Apps](https://datatracker.ietf.org/doc/html/rfc8252) in which you are using the OpenID AppAuth ([AppAuth Android](https://github.com/openid/AppAuth-Android) / [AppAuth iOS](https://github.com/openid/AppAuth-iOS)) libraries which will be using an in-app browser tab to navigate to the authorization endpoint. In which case they would be using the in-app browser for filling in the registration form.

**Comment by luis901101** [comment]
Hi, we are using [AppAuth](https://appauth.io) to SignIn/SignUp with Keycloak using Email, Google, Facebook and Apple and so far all good. But the experience in iOS is not ideal because every time the user is going to login, a mandatory iOS AlertDialog shows up indicating the app is going to access our Keycloak domain and this is not a good user experience from my point of view.
So I was wondering if there is any way to initiate an SignIn on Keycloak but handle this on App side, for instance, the app will show the Google, Facebook and Apple SignIn buttons in a native Screen so when the user taps on any of it the IDP workflow is initiated on Keycloak.

**Reply by justin-tay** [reply]
If you think the dialog impacts usability you can use the `prefersEphemeralSession` option. In which case the user doesn't get Single Sign-On (as no cookies are shared) and has to always login each time, which may also be a usability issue.

**Reply by luis901101** [reply]
Hi thanks for your feedback, yeah exactly `prefersEphemeralSession` is not good either from a usability PoV.
My idea is to create a login screen with native components complemented by Keycloak, for instance:
1. Email/Password login can be handled natively doing Oauth2 PasswordGrant.
2. Manual Sign Up not sure, I don't know if Keycloak supports registering users by some kind of endpoint POST request.
3. Social SignIn/SignUp these ones would be initiated from app I mean the call would be directly to each social login and somehow when the process completes make a request to Keycloak to indicate a social login using specific IDP was made.

The point of all of this is that the user stays within the app when they decide to SignIn/Up manually because would be the app’s own IDP, then if the user decides to SignIn/Up using Social Login then it’s expected to go outside the app interface as this is an external IDP but a known IDP I mean the user already knows and trust google.com, facebook.com but not auth.myapp.com.

**Comment by adaltonreis** [comment]
I guess the simplest approach to still comply with the patterns, would be to inform the user about the warning before the warning appears. Something like: "... you will access our servers at  APP_DOMAIN.COM/path. Please check if that is the address appearing in the address bar in the next screen".

**Comment by bhavin-android** [comment]
So , Any one find a way to use native app login form to use keycloak without use browser or in-app browser/tab. There is way we can use post api but it's use client secret key. 

**So anyone have way to use native login flow for android share it here please.**

**Reply by h-sigma** [reply]
There basically isn't a way to do this.
